crates/chain/src/keychain/txout_index.rs

```diff
@@ -99,13 +169,24 @@ const DEFAULT_LOOKAHEAD: u32 = 25;
 /// [`new`]: KeychainTxOutIndex::new
 /// [`unbounded_spk_iter`]: KeychainTxOutIndex::unbounded_spk_iter
 /// [`all_unbounded_spk_iters`]: KeychainTxOutIndex::all_unbounded_spk_iters
+// Under the hood, the KeychainTxOutIndex uses a SpkTxOutIndex that keeps track of spks, indexed by
+// descriptors. Users then assign or unassign keychains to those descriptors. It's important to
+// note that descriptors, once added, never get removed from the SpkTxOutIndex; this is useful in
+// case a user unassigns a keychain from a descriptor and after some time assigns it again.
 #[derive(Clone, Debug)]
 pub struct KeychainTxOutIndex<K> {
-    inner: SpkTxOutIndex<(K, u32)>,
-    // descriptors of each keychain
-    keychains: BTreeMap<K, Descriptor<DescriptorPublicKey>>,
+    inner: SpkTxOutIndex<(DescriptorId, u32)>,
+    // keychain -> (descriptor, descriptor id) map
+    keychains_to_descriptors: BTreeMap<K, (DescriptorId, Descriptor<DescriptorPublicKey>)>,
+    // descriptor id -> keychain map
+    descriptor_ids_to_keychain: BTreeMap<DescriptorId, (K, Descriptor<DescriptorPublicKey>)>,
```

Comment:
I think this was my suggestion. However, I just found out that this is unsound since multiple keychains can point to the same descriptor. Hence, a given `DescriptorId` can point to multiple keychains...

Reply:
This also means an outpoint can be associated with multiple keychains... Which is a bad API... The other option is to return duplicates if the caller associates the same descriptor to different keychains (which is not great either)...

Reply:
I demonstrated this problem with this test here: https://github.com/evanlinjin/bdk/commit/16366b47ef1e38fafe220433e8ad78c7ab5f4f6c

Reply:
I agree with @ValuedMammal's suggestion on discord that we should simplify this mapping for now and support more complicated scenarios later if needed.

Can we go back to a requirement that a keychain -> descriptor mapping can't change? I think this is what @danielabrozzoni originally had.  As long as we document that it's invalid/unsupported it's OK to panics if someone tries to insert a mapping for a keychain or descriptor that already exists in an existing mapping.

Reply:
@notmandatory I would like to try something before we resort to that. Give me end of today.

Reply:
I have a solution, just need to clean it up and add documentation

Reply:
I agree that we should fix the changeset not being monotonic, but I don't like the approach taken in #1428, as it's complex and feels quite hacky honestly. You're letting the user add multiple keychains for a single descriptor, but then, since the whole API is based on the assumption that there is a single keychain associated with a descriptor, you're arbitrarily deciding that the "right" descriptor is the one with the highest `Ord`. 

I would simply disallow inserting a keychain for a descriptor id that already has a keychain associated. It's the easiest solution and makes way more sense imho.

Reply:
Rather than doing something the user wouldn't expect I'd also rather disallow assigning already used descriptor_ids to new keychains. That doesn't mean we can't keep working on a more flexible solution, and by having the `descriptor_id` mapping already in this PR it sets the ground work. Further it shouldn't change the API if we can later allow changing the keychain of a descriptor, we just remove the error. 

🎉 In other news, I've tested this PR with my sqlite_store and it's working fine for my simple unit tests and with one of the wallet examples. 

🔧 A few small PRs have gone in since this PR was rebased so it probably needs to be rebased again.

Reply:
> I would simply disallow inserting a keychain for a descriptor id that already has a keychain associated. It's the easiest solution and makes way more sense imho.

If you look at `ChangeSet::keychains_added`, it allows for changing the descriptor for a keychain. If we want to implement what you suggested, the `Append::append` implementation needs to iterate through all descriptors and panic if there is a match (which would be horrible). If we only panic in `KeychainTxOutIndex`, then changesets cannot be monotonic (as shown in ffba1b56ba59ab95aa69d9f38bb790c20125d81c).

> as it's complex

I don't think it is. We are just storing multiple keychains per descriptor id. If you look at a1d55f2662b40e61c03234681be6f4c718b6b6e4, the logic barely changes and I would argue that `insert_descriptor` is now made simpler.

> arbitrarily... "right" descriptor is the one with the highest `Ord`.

Not true. It's based on `Ord` so it's not arbitrary.

> Rather than doing something the user wouldn't expect

If we document it, the caller will expect it. Same with our alternative solution (which is panicing). I would argue that the panicing approach would make our API highly unpredictable since there will be scenarios where applying an aggregated changeset will not panic, but applying the individual changesets one by one will (as mentioned above).

> Further it shouldn't change the API

It doesn't change the API at all? The behavior only defers if the caller associates the same descriptor to multiple keychains. One is to panic (which I already demonstrated is bad), the other is to use `Ord` to return the highest-ranked keychain.

---

I might need a clear argument to be convinced that the ranked-keychain solution is a bad idea. Give an example?

Reply:
> > I would simply disallow inserting a keychain for a descriptor id that already has a keychain associated. It's the easiest solution and makes way more sense imho.
> 
> If you look at `ChangeSet::keychains_added`, it allows for changing the descriptor for a keychain. If we want to implement what you suggested, the `Append::append` implementation needs to iterate through all descriptors and panic if there is a match (which would be horrible). If we only panic in `KeychainTxOutIndex`, then changesets cannot be monotonic (as shown in [ffba1b5](https://github.com/bitcoindevkit/bdk/commit/ffba1b56ba59ab95aa69d9f38bb790c20125d81c)).

Why can't the ChangeSet::append implementation be different from the insert method. The changeset can be montone while the insert method errors when you try and reassign a keychain.

> 
> > as it's complex
> 
> I don't think it is. We are just storing multiple keychains per descriptor id. If you look at [a1d55f2](https://github.com/bitcoindevkit/bdk/commit/a1d55f2662b40e61c03234681be6f4c718b6b6e4), the logic barely changes and I would argue that `insert_descriptor` is now made simpler.
> 
> > arbitrarily... "right" descriptor is the one with the highest `Ord`.
> 
> Not true. It's based on `Ord` so it's not arbitrary.
> 
> > Rather than doing something the user wouldn't expect

`Ord` of descriptors is very very arbitrary. The `cmp` output has no bearing on the semantics of the descriptor.

> 
> If we document it, the caller will expect it. Same with our alternative solution (which is panicing). I would argue that the panicing approach would make our API highly unpredictable since there will be scenarios where applying an aggregated changeset will not panic, but applying the individual changesets one by one will (as mentioned above).
> 

The documentation would be very tl;dr. It think it should just say you can't reassign keychains and we don't guarantee what happens if you do. We can then of course make whatever does happen relatively sane by maintaining montonicity in the `ChangeSet::append` method.

> > Further it shouldn't change the API
> 
> It doesn't change the API at all? The behavior only defers if the caller associates the same descriptor to multiple keychains. One is to panic (which I already demonstrated is bad), the other is to use `Ord` to return the highest-ranked keychain.
> 
> I might need a clear argument to be convinced that the ranked-keychain solution is a bad idea. Give an example?

The code is just complicated and the documentation is long winded and it doesn't get us anything. Sure making the ChangeSet monotone is a good idea but functions called on the actual txout index can just panic or error when you try to reassign keychains.

Reply:
Just had a call with @LLFourn. He pointed out that my changes in #1428 does not actually make changesets monotone as we still need to rely on the order of the changesets. He also noted that there may be a misunderstanding that the ranking is based on `Ord` of descriptors when this is not the case. We are basing it on the `Ord` of keychains. He also noted that the main value of these changes is to allow multiple keychains to point to the same descriptor (since we can avoid doing #1390). He also noted that when he proposed the `DescriptorId` concept, he didn't take into consideration that we can just disallow changing the descriptor associated with a keychain once it is set in `Append::append` (basically ignoring reassignment). If we disallow reassigning descriptors, we can do away with the `DescriptorId` completely.

Reply:
In either of these solutions, @LLFourn pointed out that we should error when we call `insert_descriptor` when trying to reassign a keychain's descriptor. However, we should either ignore/allow it (no error/panic) when calling `apply_changeset` or `Append::append`.

Reply:
> He pointed out that my changes in https://github.com/bitcoindevkit/bdk/pull/1428 does not actually make changesets monotone as we still need to rely on the order of the changesets.

Why is it a problem if applying changesets is not commutative?

The way I see it, we have a KeychainTxOutIndex with monotone descriptors, and these descriptors can be labeled with certain keychains, and labels can be removed. One operation we can do on the index is "add this certain label for this descriptor", reflected in the `keychains_added` map. If we want this "relabeling" operation to still exist, then I don't see a way for having commutative changesets. Labeling a descriptor with `K` and then relabeling it with `K'` is always going to yield a different result than labeling it with `K'` and then with `K`.

Reply:
It's not a problem. I was just using the wrong terminology.